src/main/java/org/traccar/helper/CoordinateUtil.java

```diff
@@ -0,0 +1,247 @@
+     * @return Whether the coordinates are in non-China area
+     */
+    public static boolean outOfChina(double longitude, double latitude) {
+        return (longitude < 72.004 || longitude > 137.8347) || (latitude < 0.8293 || latitude > 55.8271);
+    }
+
+    /**
+     * Convert WGS84 to GCJ-02
+     *
+     * @param longitude Longitude
+     * @param latitude Latitude
+     * @return GCJ-02 coordinate
+     */
+    public static Coordinate wgs84ToGcj02(double longitude, double latitude) {
+        return new Coordinate(longitude, latitude).offset(offset(longitude, latitude, true));
+    }
+
+    /**
+     * Convert GCJ-02 to WGS84
+     *
+     * @param longitude Longitude
+     * @param latitude Latitude
+     * @return WGS84 coordinate
+     */
+    public static Coordinate gcj02ToWgs84(double longitude, double latitude) {
+        return new Coordinate(longitude, latitude).offset(offset(longitude, latitude, false));
+    }
+
+    /**
+     * Offset algorithm for WGS84 to GCJ-02 conversion (not exact)
+     *
+     * @param longitude Longitude
+     * @param latitude Latitude
+     * @param isPlus Forward offset or not: WGS84 to GCJ-02 use forward, otherwise use reverse.
+     * @return Offset coordinates
+     */
+    private static Coordinate offset(double longitude, double latitude, boolean isPlus) {
+        double dlng = transLng(longitude - 105.0, latitude - 35.0);
+        double dlat = transLat(longitude - 105.0, latitude - 35.0);
+
+        double magic = Math.sin(latitude / 180.0 * PI);
+        magic = 1 - CORRECTION_PARAM * magic * magic;
+        final double sqrtMagic = Math.sqrt(magic);
+
+        dlng = (dlng * 180.0) / (RADIUS / sqrtMagic * Math.cos(latitude / 180.0 * PI) * PI);
+        dlat = (dlat * 180.0) / ((RADIUS * (1 - CORRECTION_PARAM)) / (magic * sqrtMagic) * PI);
+
+        if (!isPlus) {
+            dlng = -dlng;
+            dlat = -dlat;
+        }
+
+        return new Coordinate(dlng, dlat);
+    }
+
+    /**
+     * Calculate longitude coordinates
+     *
+     * @param longitude Longitude
+     * @param latitude Latitude
+     * @return ret Calculated
+     */
+    private static double transLng(double longitude, double latitude) {
+        double ret = 300.0 + longitude + 2.0 * latitude
+                + 0.1 * longitude * longitude + 0.1 * longitude * latitude + 0.1 * Math.sqrt(Math.abs(longitude));
+        ret += (20.0 * Math.sin(6.0 * longitude * PI) + 20.0 * Math.sin(2.0 * longitude * PI)) * 2.0 / 3.0;
+        ret += (20.0 * Math.sin(longitude * PI) + 40.0 * Math.sin(longitude / 3.0 * PI)) * 2.0 / 3.0;
+        ret += (150.0 * Math.sin(longitude / 12.0 * PI) + 300.0 * Math.sin(longitude / 30.0 * PI)) * 2.0 / 3.0;
+        return ret;
+    }
+
+    /**
+     * Calculate latitude coordinates
+     *
+     * @param longitude Longitude
+     * @param latitude Latitude
+     * @return ret Calculated
+     */
+    private static double transLat(double longitude, double latitude) {
```

Comment:
Did you copy all this code from somewhere?

Reply:
Yes, most of the code for this class is taken from "https://github.com/dromara/hutool/blob/v5-master/hutool-core/src/main/java/cn/hutool/core/util/ CoordinateUtil.java", the tool library from which it was copied. There are a few major modifications made:
1. I translated the Chinese comments in the original class;
2. I removed the code about BD09 coordinate system. This coordinate system is only used in Baidu map.
3. The code uses a lot of "lat" and "lng" abbreviations to express latitude and longitude, which is not consistent with the original style of the project, so I modified these abbreviations.

Reply:
Yes, most of the code for this class is taken from "https://github.com/dromara/hutool/blob/v5-master/hutool-core/src/main/java/cn/hutool/core/util/CoordinateUtil.java", the tool library from which it was copied. There are a few major modifications made:
1. I translated the Chinese comments in the original class;
2. I removed the code about BD09 coordinate system. This coordinate system is only used in Baidu map.
3. The code uses a lot of "lat" and "lng" abbreviations to express latitude and longitude, which is not consistent with the original style of the project, so I modified these abbreviations.

Reply:
Basically we need to update this code to match our style. We only use full words, no shortened versions.

Reply:
OK, I'll modify this part of the code. Also I will adjust the order of the parameters, because the "hutool" library is coded by Chinese. In China, when we use latitude and longitude to express geographic coordinates, we usually write longitude first, then latitude, but in English, we usually write latitude first, then longitude. So I will also change this to unify the style of the code.